src/System.Net.Http/tests/UnitTests/HPack/HPackRoundtripTests.cs

```diff
@@ -0,0 +1,214 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http.HPack;
+using System.Net.Http.Headers;
+using System.Text;
+using Xunit;
+
+namespace System.Net.Http.Unit.Tests.HPack
+{
+    public class HPackRoundtripTests
+    {
+
+        #region "tests"
+
+        // TODO: Extend to Property-Based Test using FsCheck 
+
+        [Fact]
+        public void HPack_Roundtrip_Single_Header_With_Single_Value()
+        {
+            HttpHeaders headers =
+                BuildHttpHeaders(
+                    new[] {
+                        ("header", new[] { "value" })
+                    }
+                );
+
+            Memory<byte> encoding = HPackEncode(headers);
+            HttpHeaders decodedHeaders = HPackDecode(encoding);
+
+            CompareHttpHeaders(headers, decodedHeaders);
+        }
+
+        [Fact]
+        public void HPack_Roundtrip_Single_Header_With_Two_Values()
+        {
+            HttpHeaders headers =
+                BuildHttpHeaders(
+                    new[] {
+                        ("header", new[] { "value1", "value2" })
+                    }
+                );
+
+            Memory<byte> encoding = HPackEncode(headers);
+            HttpHeaders decodedHeaders = HPackDecode(encoding);
+
+            CompareHttpHeaders(headers, decodedHeaders);
+        }
+
+        [Fact]
+        public void HPack_Roundtrip_Multiple_Headers()
+        {
+            HttpHeaders headers =
+                BuildHttpHeaders(
+                    new[] {
+                        ("header-0", new[] { "value1", "value2" }),
+                        ("header-0", new[] { "value3" }),
+                        ("header-1", new[] { "value1" }),
+                        ("header-2", new[] { "value1", "value2" })
+                    }
+                );
+
+            Memory<byte> encoding = HPackEncode(headers);
+            HttpHeaders decodedHeaders = HPackDecode(encoding);
+
+            CompareHttpHeaders(headers, decodedHeaders);
+        }
+
+        #endregion
+
+        #region "helpers"
+
+        private HttpHeaders BuildHttpHeaders((string key, string[] values)[] seedValues)
+        {
+            var headers = new HttpRequestHeaders();
+
+            foreach ((string key, string[] value) header in seedValues)
+            {
+                headers.Add(header.key, header.value);
+            }
+
+            return headers;
+        }
+
+        private void CompareHttpHeaders(HttpHeaders expected, HttpHeaders actual)
+        {
+            Assert.Equal(expected.Count(), actual.Count());
+
+            expected
+                .Zip(actual)
+                .ToList()
+                .ForEach(x => CompareHttpHeader(x.First, x.Second));
```

Comment:
Seems like using Assert.ForAll would be better.

Reply:
Done